Image feed/sprint 10

alekseiypetrov opened the pull request.

Commit:
Добавлены 2 перечисления: Constants (хранение необходимых данных для API) и Colors (раскраска UI).

Commit:
Из Constants убран enum Colors, все цвета были перенесены в Assets.
Выполнена верстка двух экранов: Auth и WebView.

Commit:
На экране WebView добавлен Progress Bar для визуализации прогресса загрузки страницы при переходе на экран. Также добавлена технология KVO к Progress Bar, чтобы пользователю было видно, на сколько загружена веб-страница.
Добавлено событие для обработки нажатия кнопки "Назад" в WebView.

Commit:
Добавлен запрос для получения токена.

Commit:
Доделана логика работы экрана авторизации и добавлен главный экран - SplashViewController.

Kirchberg reviewed (changes requested):
⚠️ Критические замечания (это нужно исправить обязательно):

1. Приложение падает с ошибкой при попытке нажатия на кнопку войти – Exception	NSException *	"Could not instantiate class named WKWebView because no class named WKWebView was found; the class needs to be defined in source code or linked in from a library (ensure the class is part of the correct target)"	0x0000600000c8a670. (Нарушены пункты 3, 6.2, 6.3, [см. скриншот](https://ibb.co/HLvjqQJr))
2. В проекте есть несколько мест, где используется принудительное извлечение опционалов. В Constants.swift константа defaultBaseURL создается с !. Это небезопасно и может привести к падению приложения. Следует использовать безопасное извлечение. (Нарушены пункты 5, 7.4)
3. В файле WebViewViewController.swift в методе code(from:) допущена опечатка в пути, по которому проверяется URL для получения кода авторизации: urlComponents.path == "/ouath/authrize/native". Правильный путь должен быть /oauth/authorize/native. Из-за этой ошибки приложение никогда не сможет успешно получить код и завершить авторизацию. (Нарушен пункт 6.4)
4. В OAuth2Service.swift функция makeOAuthTokenRequest в случае ошибки создания URL-компонентов или URL просто возвращает nil. Вызывающий код в fetchOAuthToken также молча завершает работу. Это нарушает требование логирования ошибок при создании запроса и затрудняет отладку. (Нарушен пункт 13.1)

Без исправлений этих пунктов, я не смогу принять у тебя работу. Если возникнут вопросы, обязательно задавай их Наставнику!

--- 

🍏 Рекомендации (это можно исправить по желанию):

1. В расширении URLSession+data.swift есть логическая ошибка. После успешного получения данных (if let data = data), вызывается fulfillCompletionOnTheMainThread(.success(data)), но выполнение функции не прерывается. Код продолжает выполняться и доходит до fulfillCompletionOnTheMainThread(.failure(NetworkError.urlSessionError)), вызывая completion-блок второй раз. Чтобы это исправить, нужно добавить return после вызова .success.

--- 

🍪 Для размышлений 

Для того, чтобы не работать с `snake_case` ключами, не описывать их в `CodingKeys`, ты можешь также 
настраивать `JSONDecoder` под тот вид данных, который приходит от сервера. 
Рекомендую изучить этот способ самостоятельно https://nemecek.be/blog/90/how-to-decode-snake-case-with-codable

---

Чтобы твоя работа была зачтена, необходимо исправить все критические замечания.
Жду исправлений :)

Commit:
Устранены недостатки , указанные в последнем ревью.

Commit:
Поправил настройки проекта:
1) версия IOS. Почему-то минимальный iOS стоял 17.0, а не 13.0)
2) добавил WebKit.framework в раздел Frameworks, Libraries

Commit:
Поправил недочет с датой (так как Date.now появился с iOS 15.0 и новее, а до этого его не было) и различные кнопки в сториборде.

alekseiypetrov commented:
Обращение к ревьюеру Кириллу.
Кирилл, я, вообще говоря, поправил настройки проекта (можно посмотреть по предпоследнему комиту). Если и сейчас не заработает, то, если честно, я бессилен в данной ситуации, потому что у меня не крашилось приложение ни в симуляторе, ни на живом айфоне даже без этих настроек. Буду надеяться, что все заработает как надо.

Kirchberg commented:
> Обращение к ревьюеру Кириллу. Кирилл, я, вообще говоря, поправил настройки проекта (можно посмотреть по предпоследнему комиту). Если и сейчас не заработает, то, если честно, я бессилен в данной ситуации, потому что у меня не крашилось приложение ни в симуляторе, ни на живом айфоне даже без этих настроек. Буду надеяться, что все заработает как надо.

Привет, Алексей, все в порядке, последние 2 коммита помогли приложению не падать, все хорошо :)

Не забудь вмерджить пул-реквест в основную ветку.
Успехов в следующем задании!
